File: api/app/simulation.py
"""Deterministic Phase 6 simulation clock, claw snapping and movement samples.

Nothing in this module reads the wall clock. The simulation clock is a whole number of
bounded ticks, and a vehicle position is a pure function of the published route plan
plus that clock. A tick never carries a scenario revision, so advancing the clock can
never overwrite a newer revision: the revision only changes when a command mutates the
scenario, and a tick that was computed for an older revision is simply discarded by the
client.

Units stay explicit: the clock is seconds, positions are metres in the local Three.js
XZ plane, and headings are compass-style degrees (0 points to +z, 90 points to +x).
"""

from __future__ import annotations

from dataclasses import dataclass
from math import atan2, degrees, hypot
from typing import Any, Final

from .routing import travel_seconds

# One bounded simulation tick. The clock advances in whole ticks, so two clients that
# apply the same number of ticks always publish the same elapsed time and the same
# vehicle positions.
TICK_SECONDS: Final = 0.5
# Hard bound on a single clock advance: a backgrounded tab cannot jump the animation
# past the bounded window when it resumes.
MAX_TICKS_PER_ADVANCE: Final = 240

# Bounds for `simulation.speedMultiplier`, matching the frozen scenario schema.
MIN_SPEED_MULTIPLIER: Final = 0.25
MAX_SPEED_MULTIPLIER: Final = 8.0
DEFAULT_SPEED_MULTIPLIER: Final = 1.0

# Claw snap radius; mirrors `SNAP_NODE_MAX_RADIUS_M` in `frontend/src/city/dataset.ts`.
SNAP_NODE_MAX_RADIUS_M: Final = 12.0
EPSILON_M: Final = 1e-9


@dataclass(frozen=True, slots=True)
class NodeSnap:
    """Closest road node inside the claw radius."""

    node_id: str
    kind: str
    position: dict[str, float]
    distance_meters: float


@dataclass(frozen=True, slots=True)
class RouteLeg:
    """One driveable edge of a vehicle route, with its whole-second travel time."""

    edge_id: str
    from_node_id: str
    to_node_id: str
    length_meters: float
    seconds: int


@dataclass(frozen=True, slots=True)
class VehicleSample:
    """Where one vehicle is after a whole number of ticks."""

    vehicle_id: str
    node_id: str
    edge_id: str | None
    progress: float
    position: dict[str, float]
    heading_degrees: float
    travelled_seconds: float
    arrived: bool


def clamp_speed_multiplier(value: float) -> float:
    """Validate and round one simulation speed multiplier.

    The frozen schema allows any positive multiplier up to eight. Values outside the
    range are rejected instead of clamped, so a bad request never looks accepted.
    """
    multiplier = float(value)
    if multiplier <= 0 or multiplier > MAX_SPEED_MULTIPLIER:
        raise ValueError(
            f"speedMultiplier must be greater than 0 and at most {MAX_SPEED_MULTIPLIER:g}"
        )
    return round(multiplier, 3)


def simulation_state(
    *, running: bool, speed_multiplier: float = DEFAULT_SPEED_MULTIPLIER, tick: int = 0
) -> dict[str, Any]:
    """Build the frozen ``simulation`` block for one revision.

    ``elapsedSeconds`` is derived from ``tick``, so the two fields can never disagree.
    """
    bounded_tick = max(0, int(tick))
    return {
        "running": bool(running),
        "speedMultiplier": round(float(speed_multiplier), 3),
        "tick": bounded_tick,
        "elapsedSeconds": round(bounded_tick * TICK_SECONDS, 3),
    }


def bounded_tick_delta(real_seconds: float, speed_multiplier: float) -> int:
    """Whole ticks earned by ``real_seconds`` of real time at the current speed.

    The result is bounded by ``MAX_TICKS_PER_ADVANCE`` and never negative, so a stalled
    or resumed tab produces a bounded, deterministic step instead of a jump.
    """
    if real_seconds <= 0:
        return 0
    multiplier = float(speed_multiplier)
    if multiplier <= 0:
        return 0
    return min(MAX_TICKS_PER_ADVANCE, int(real_seconds * multiplier / TICK_SECONDS))


def advance_clock(simulation: dict[str, Any], ticks: int) -> dict[str, Any]:
    """Advance one simulation block by a bounded, non-negative number of ticks."""
    bounded_ticks = max(0, min(int(ticks), MAX_TICKS_PER_ADVANCE))
    return simulation_state(
        running=bool(simulation.get("running", False)),
        speed_multiplier=float(simulation.get("speedMultiplier", DEFAULT_SPEED_MULTIPLIER)),
        tick=int(simulation.get("tick", 0)) + bounded_ticks,
    )


def nearest_node(
    graph: dict[str, Any],
    point: dict[str, float],
    max_radius: float = SNAP_NODE_MAX_RADIUS_M,
) -> NodeSnap | None:
    """Closest road node inside ``max_radius``, or ``None`` outside it.

    An exact tie (within ``EPSILON_M``) always goes to the lexicographically smaller
    node id, so the same drop resolves to the same node on every run.
    """
    best: NodeSnap | None = None
    for node in graph["nodes"]:
        position = node["position"]
        distance = hypot(float(point["x"]) - position["x"], float(point["z"]) - position["z"])
        if distance > max_radius + EPSILON_M:
            continue
        if best is None or distance < best.distance_meters - EPSILON_M:
            best = NodeSnap(node["nodeId"], node["kind"], position, distance)
            continue
        is_tie = abs(distance - best.distance_meters) <= EPSILON_M
        if is_tie and node["nodeId"] < best.node_id:
            best = NodeSnap(node["nodeId"], node["kind"], position, distance)
    return best


def heading_degrees(from_position: dict[str, float], to_position: dict[str, float]) -> float:
    """Compass-style heading of a direction: 0 points to +z, 90 points to +x."""
    dx = to_position["x"] - from_position["x"]
    dz = to_position["z"] - from_position["z"]
    return (degrees(atan2(dx, dz)) + 360) % 360


def route_legs(
    graph: dict[str, Any],
    node_sequence: list[str],
    edge_sequence: list[str],
    vehicle_speed_kph: float | None = None,
) -> list[RouteLeg]:
    """Ordered, driveable legs of a published route.

    A sequence that is not connected stops there instead of inventing a shortcut, and a
    leg's travel time is ``ceil(length / min(road limit, vehicle speed))`` exactly like
    the Phase 5 planner, so the animation cannot outrun the plan it renders.
    """
    if not edge_sequence:
        return []
    nodes = {node["nodeId"]: node for node in graph["nodes"]}
    edges = {edge["edgeId"]: edge for edge in graph["edges"]}
    legs: list[RouteLeg] = []
    cursor = node_sequence[0] if node_sequence else None
    for edge_id in edge_sequence:
        edge = edges.get(edge_id)
        if edge is None or cursor is None:
            break
        if edge["fromNodeId"] == cursor:
            origin, destination = edge["fromNodeId"], edge["toNodeId"]
        elif edge["toNodeId"] == cursor:
            origin, destination = edge["toNodeId"], edge["fromNodeId"]
        else:
            break
        length = float(edge["lengthMeters"])
        legs.append(
            RouteLeg(
                edge_id=edge_id,
                from_node_id=origin,
                to_node_id=destination,
                length_meters=length,
                seconds=travel_seconds(length, float(edge["speedLimitKph"]), vehicle_speed_kph),
            )
        )
        cursor = destination
    return legs


def _sample_legs(
    graph: dict[str, Any],
    legs: list[RouteLeg],
    elapsed_seconds: float,
    start_node_id: str,
) -> VehicleSample:
    """Interpolate one position along the route at ``elapsed_seconds``.

    The sample is clamped to the end of the route: a route that has ended stays at its
    last stop instead of wrapping around.
    """
    positions = {node["nodeId"]: node["position"] for node in graph["nodes"]}
    start_position = positions[start_node_id]
    if not legs:
        return VehicleSample(
            vehicle_id="",
            node_id=start_node_id,
            edge_id=None,
            progress=0.0,
            position={"x": start_position["x"], "y": start_position["y"], "z": start_position["z"]},
            heading_degrees=0.0,
            travelled_seconds=0.0,
            arrived=True,
        )

    total = sum(leg.seconds for leg in legs)
    clamped = max(0.0, float(elapsed_seconds))
    remaining = clamped
    for index, leg in enumerate(legs):
        is_last = index == len(legs) - 1
        if remaining < leg.seconds or is_last:
            progress = 1.0 if leg.seconds <= 0 else min(1.0, max(0.0, remaining / leg.seconds))
            from_position = positions[leg.from_node_id]
            to_position = positions[leg.to_node_id]
            position = {
                "x": from_position["x"] + (to_position["x"] - from_position["x"]) * progress,
                "y": from_position["y"] + (to_position["y"] - from_position["y"]) * progress,
                "z": from_position["z"] + (to_position["z"] - from_position["z"]) * progress,
            }
            arrived = clamped >= total
            if arrived:
                progress = 1.0
                position = {
                    "x": to_position["x"],
                    "y": to_position["y"],
                    "z": to_position["z"],
                }
            return VehicleSample(
                vehicle_id="",
                # Mid-edge the vehicle still occupies the node it departed; once it has
                # arrived it occupies the node it reached.
                node_id=leg.to_node_id if arrived else leg.from_node_id,
                edge_id=None if arrived else leg.edge_id,
                progress=progress,
                position=position,
                heading_degrees=heading_degrees(from_position, to_position),
                travelled_seconds=min(clamped, float(total)),
                arrived=arrived,
            )
        remaining -= leg.seconds

    final_leg = legs[-1]
    final_position = positions[final_leg.to_node_id]
    return VehicleSample(
        vehicle_id="",
        node_id=final_leg.to_node_id,
        edge_id=None,
        progress=1.0,
        position={"x": final_position["x"], "y": final_position["y"], "z": final_position["z"]},
        heading_degrees=heading_degrees(positions[final_leg.from_node_id], final_position),
        travelled_seconds=float(total),
        arrived=True,
    )


def vehicle_samples(
    graph: dict[str, Any],
    route_plan: dict[str, Any] | None,
    vehicles: list[dict[str, Any]],
    elapsed_seconds: float,
) -> list[VehicleSample]:
    """Deterministic sample for every vehicle at ``elapsed_seconds``.

    A vehicle without a published route, or with a route without stops, stays on its
    current node. Samples are ordered by vehicle id so the result is stable.
    """
    routes = {
        route["vehicleId"]: route
        for route in (route_plan or {}).get("vehicles", [])
    }
    samples: list[VehicleSample] = []
    for vehicle in sorted(vehicles, key=lambda item: item["vehicleId"]):
        vehicle_id = vehicle["vehicleId"]
        route = routes.get(vehicle_id)
        start_node = vehicle.get("currentNodeId") or ""
        speed = float(vehicle.get("speedKilometersPerHour", 0)) or None
        legs = (
            route_legs(
                graph,
                route.get("nodeSequence", []),
                route.get("edgeSequence", []),
                speed,
            )
            if route and route.get("stops")
            else []
        )
        sample = _sample_legs(
            graph,
            legs,
            elapsed_seconds,
            start_node,
        )
        samples.append(
            VehicleSample(
                vehicle_id=vehicle_id,
                node_id=sample.node_id,
                edge_id=sample.edge_id,
                progress=sample.progress,
                position=sample.position,
                heading_degrees=sample.heading_degrees,
                travelled_seconds=sample.travelled_seconds,
                arrived=sample.arrived,
            )
        )
    return samples

File: frontend/src/scenario/simulation.ts
/**
 * Phase 6 simulation clock and movement, as pure functions.
 *
 * The backend owns the authoritative clock and publishes it in every snapshot. This
 * module applies the same bounded rule to the browser animation, so a vehicle position
 * is a pure function of the published route plan plus the tick clock: the same clock
 * always produces the same position, and two clients given the same snapshot agree.
 *
 * Everything here is independent of WebGL on purpose, which is what keeps the movement
 * unit-testable.
 */

import { createRoadNetwork, type CityPoint, type RoadNetwork } from '../city/dataset';

import type { RoutePlan, ScenarioSnapshot, SimulationState, Vehicle, VehicleRoute } from './scenario';

/** One bounded simulation tick: the clock only moves in whole ticks. */
export const SIMULATION_TICK_SECONDS = 0.5;
/** Hard bound on one clock advance, so a stalled tab cannot jump the animation. */
export const MAX_TICKS_PER_ADVANCE = 240;
export const MIN_SIMULATION_SPEED = 0.25;
export const MAX_SIMULATION_SPEED = 8;
export const DEFAULT_SIMULATION_SPEED = 1;
/** The speeds the control offers. Every one of them is inside the frozen bounds. */
export const SIMULATION_SPEED_CHOICES = [0.5, 1, 2, 4, 8] as const;

export interface RouteLeg {
  edgeId: string;
  fromNodeId: string;
  toNodeId: string;
  lengthMeters: number;
  /** Whole seconds at the slower of the road limit and the robot speed. */
  seconds: number;
}

export interface VehicleSample {
  vehicleId: string;
  /** The node the vehicle occupies: where it departed, or where it arrived. */
  nodeId: string;
  edgeId: string | null;
  progress: number;
  position: CityPoint;
  headingDegrees: number;
  travelledSeconds: number;
  arrived: boolean;
}

export function clampSpeedMultiplier(multiplier: number): number {
  const value = Number(multiplier);
  if (!Number.isFinite(value) || value <= 0 || value > MAX_SIMULATION_SPEED) {
    throw new Error(
      `speedMultiplier must be greater than 0 and at most ${MAX_SIMULATION_SPEED}`,
    );
  }
  return Math.round(value * 1000) / 1000;
}

export function simulationState(options: {
  running: boolean;
  speedMultiplier?: number;
  tick?: number;
}): SimulationState {
  const tick = Math.max(0, Math.trunc(options.tick ?? 0));
  return {
    running: Boolean(options.running),
    speedMultiplier: Math.round((options.speedMultiplier ?? DEFAULT_SIMULATION_SPEED) * 1000) / 1000,
    tick,
    elapsedSeconds: Math.round(tick * SIMULATION_TICK_SECONDS * 1000) / 1000,
  };
}

/** Whole ticks earned by `realSeconds` at `speedMultiplier`, bounded and non-negative. */
export function boundedTickDelta(realSeconds: number, speedMultiplier: number): number {
  if (!(realSeconds > 0) || !(speedMultiplier > 0)) return 0;
  return Math.min(
    MAX_TICKS_PER_ADVANCE,
    Math.trunc((realSeconds * speedMultiplier) / SIMULATION_TICK_SECONDS),
  );
}

/** Advance one clock by a bounded, non-negative number of ticks. */
export function advanceSimulationClock(
  simulation: SimulationState,
  ticks: number,
): SimulationState {
  const bounded = Math.max(0, Math.min(Math.trunc(ticks), MAX_TICKS_PER_ADVANCE));
  return simulationState({
    running: simulation.running,
    speedMultiplier: simulation.speedMultiplier,
    tick: simulation.tick + bounded,
  });
}

export function travelSeconds(
  lengthMeters: number,
  speedLimitKph: number,
  vehicleSpeedKph: number,
): number {
  const effectiveKph = Math.min(speedLimitKph, vehicleSpeedKph);
  if (!(effectiveKph > 0)) throw new Error('effective speed must be positive');
  return Math.ceil(lengthMeters / ((effectiveKph * 1000) / 3600));
}

/**
 * Ordered, driveable legs of a published route.
 *
 * A sequence that is not connected stops there instead of inventing a shortcut, and a
 * leg uses the planner's whole-second rule, so the animation cannot outrun the plan it
 * renders.
 */
export function routeLegs(
  network: RoadNetwork,
  nodeSequence: readonly string[],
  edgeSequence: readonly string[],
  vehicleSpeedKph: number,
): RouteLeg[] {
  if (edgeSequence.length === 0) return [];
  const legs: RouteLeg[] = [];
  let cursor: string | null = nodeSequence[0] ?? null;
  for (const edgeId of edgeSequence) {
    const edge = network.edges.get(edgeId);
    if (!edge || cursor === null) break;
    let origin: string;
    let destination: string;
    if (edge.fromNodeId === cursor) {
      origin = edge.fromNodeId;
      destination = edge.toNodeId;
    } else if (edge.toNodeId === cursor) {
      origin = edge.toNodeId;
      destination = edge.fromNodeId;
    } else {
      break;
    }
    legs.push({
      edgeId,
      fromNodeId: origin,
      toNodeId: destination,
      lengthMeters: edge.lengthMeters,
      seconds: travelSeconds(edge.lengthMeters, edge.speedLimitKph, vehicleSpeedKph),
    });
    cursor = destination;
  }
  return legs;
}

export function headingDegrees(from: CityPoint, to: CityPoint): number {
  const degrees = (Math.atan2(to.x - from.x, to.z - from.z) * 180) / Math.PI;
  return (degrees + 360) % 360;
}

/** Interpolate one route position at `elapsedSeconds`, clamped to the last stop. */
export function sampleVehicleAt(
  network: RoadNetwork,
  legs: readonly RouteLeg[],
  elapsedSeconds: number,
  startNodeId: string,
): VehicleSample {
  const startPosition = nodePosition(network, startNodeId);
  if (legs.length === 0) {
    return {
      vehicleId: '',
      nodeId: startNodeId,
      edgeId: null,
      progress: 0,
      position: startPosition,
      headingDegrees: 0,
      travelledSeconds: 0,
      arrived: true,
    };
  }

  const total = legs.reduce((sum, leg) => sum + leg.seconds, 0);
  const clamped = Math.max(0, elapsedSeconds);
  let remaining = clamped;
  for (let index = 0; index < legs.length; index += 1) {
    const leg = legs[index];
    const isLast = index === legs.length - 1;
    if (remaining < leg.seconds || isLast) {
      const arrived = clamped >= total;
      const progress = arrived
        ? 1
        : leg.seconds <= 0
          ? 1
          : Math.min(1, Math.max(0, remaining / leg.seconds));
      const from = nodePosition(network, leg.fromNodeId);
      const to = nodePosition(network, leg.toNodeId);
      const position = arrived
        ? to
        : {
            x: from.x + (to.x - from.x) * progress,
            y: from.y + (to.y - from.y) * progress,
            z: from.z + (to.z - from.z) * progress,
          };
      return {
        vehicleId: '',
        nodeId: arrived ? leg.toNodeId : leg.fromNodeId,
        edgeId: arrived ? null : leg.edgeId,
        progress,
        position,
        headingDegrees: headingDegrees(from, to),
        travelledSeconds: Math.min(clamped, total),
        arrived,
      };
    }
    remaining -= leg.seconds;
  }

  const last = legs[legs.length - 1];
  const lastPosition = nodePosition(network, last.toNodeId);
  return {
    vehicleId: '',
    nodeId: last.toNodeId,
    edgeId: null,
    progress: 1,
    position: lastPosition,
    headingDegrees: headingDegrees(nodePosition(network, last.fromNodeId), lastPosition),
    travelledSeconds: total,
    arrived: true,
  };
}

function nodePosition(network: RoadNetwork, nodeId: string): CityPoint {
  const node = network.nodes.get(nodeId);
  if (node) return node.position;
  return { x: 0, y: 0, z: 0 };
}

/**
 * Deterministic sample for every vehicle at `elapsedSeconds`.
 *
 * An empty array means "do not move anything": there is no route plan, or the plan
 * belongs to an older revision than the snapshot. A vehicle with no stops stays on its
 * current node.
 */
export function vehicleSamples(
  snapshot: ScenarioSnapshot,
  elapsedSeconds: number,
  network: RoadNetwork = createRoadNetwork(),
): VehicleSample[] {
  const plan = currentRoutePlan(snapshot);
  if (!plan) return [];
  const routes = new Map(plan.vehicles.map((route) => [route.vehicleId, route]));
  return [...snapshot.vehicles]
    .sort((left, right) => (left.vehicleId < right.vehicleId ? -1 : 1))
    .map((vehicle) => {
      const route = routes.get(vehicle.vehicleId);
      const legs =
        route && route.stops.length > 0
          ? routeLegs(
              network,
              route.nodeSequence,
              route.edgeSequence,
              vehicle.speedKilometersPerHour,
            )
          : [];
      const sample = sampleVehicleAt(
        network,
        legs,
        elapsedSeconds,
        vehicle.currentNodeId ?? '',
      );
      return { ...sample, vehicleId: vehicle.vehicleId };
    });
}

/** The route plan only counts while it belongs to the current revision. */
export function currentRoutePlan(snapshot: ScenarioSnapshot): RoutePlan | null {
  const plan = snapshot.routePlan;
  if (!plan || plan.scenarioRevision !== snapshot.scenarioRevision) return null;
  return plan;
}

export function routeForVehicle(plan: RoutePlan, vehicleId: string): VehicleRoute | null {
  return plan.vehicles.find((route) => route.vehicleId === vehicleId) ?? null;
}

/** Route surfaces for the current plan: one per vehicle that actually drives. */
export function routeSurfaces(
  snapshot: ScenarioSnapshot,
): { vehicleId: string; startNodeId: string | null; edgeIds: readonly string[] }[] {
  const plan = currentRoutePlan(snapshot);
  if (!plan) return [];
  return plan.vehicles
    .filter((route) => route.edgeSequence.length > 0)
    .map((route) => ({
      vehicleId: route.vehicleId,
      startNodeId: route.nodeSequence[0] ?? null,
      edgeIds: route.edgeSequence,
    }));
}

/**
 * Scene placement for every vehicle at `elapsedSeconds`.
 *
 * These are plain values, not Three.js objects, so the mapping from the snapshot to the
 * scene stays testable without a renderer.
 */
export function vehiclePlacements(
  snapshot: ScenarioSnapshot,
  elapsedSeconds: number,
  liftedVehicleId: string | null = null,
  network: RoadNetwork = createRoadNetwork(),
): {
  vehicleId: string;
  nodeId: string;
  position: CityPoint;
  headingDegrees: number;
  lifted: boolean;
}[] {
  const samples = vehicleSamples(snapshot, elapsedSeconds, network);
  if (samples.length > 0) {
    return samples.map((sample) => ({
      vehicleId: sample.vehicleId,
      nodeId: sample.nodeId,
      position: sample.position,
      headingDegrees: sample.headingDegrees,
      lifted: sample.vehicleId === liftedVehicleId,
    }));
  }
  // No plan yet: the fleet still has to be visible, parked on its current node.
  return [...snapshot.vehicles]
    .sort((left, right) => (left.vehicleId < right.vehicleId ? -1 : 1))
    .map((vehicle) => ({
      vehicleId: vehicle.vehicleId,
      nodeId: vehicle.currentNodeId ?? '',
      position: nodePosition(network, vehicle.currentNodeId ?? ''),
      headingDegrees: 0,
      lifted: vehicle.vehicleId === liftedVehicleId,
    }));
}

export function isSimulationRunning(snapshot: ScenarioSnapshot | null): boolean {
  return Boolean(snapshot?.simulation.running);
}

/** Human-readable clock for the control panel. */
export function simulationClockLabel(simulation: SimulationState): string {
  const seconds = simulation.elapsedSeconds;
  const state = simulation.running ? 'Running' : simulation.tick > 0 ? 'Paused' : 'Stopped';
  return `${state} · tick ${simulation.tick} · ${seconds.toFixed(1)}s · x${simulation.speedMultiplier}`;
}

export function isSpeedChoice(value: number): boolean {
  return (SIMULATION_SPEED_CHOICES as readonly number[]).includes(value);
}

export function vehicleById(
  snapshot: ScenarioSnapshot | null,
  vehicleId: string | null,
): Vehicle | null {
  if (!snapshot || !vehicleId) return null;
  return snapshot.vehicles.find((vehicle) => vehicle.vehicleId === vehicleId) ?? null;
}
